File: modules/ai/router.go
package ai

// ---------------------------------------------------------------------------
// ModelRouter
// ---------------------------------------------------------------------------

// ModelCategory classifies the type of AI task being performed.
type ModelCategory int

const (
	CategoryCoding   ModelCategory = iota // code generation/editing
	CategoryPlanning                       // task planning and decomposition
	CategoryReview                         // code review and analysis
	CategoryAnswer                         // question answering
	CategoryExplore                        // exploration and research
	CategoryCreative                       // creative writing
	CategoryAnalysis                       // data analysis
	CategoryGeneral                        // general-purpose tasks
)

// String returns the human-readable name of the model category.
func (c ModelCategory) String() string {
	switch c {
	case CategoryCoding:
		return "coding"
	case CategoryPlanning:
		return "planning"
	case CategoryReview:
		return "review"
	case CategoryAnswer:
		return "answer"
	case CategoryExplore:
		return "explore"
	case CategoryCreative:
		return "creative"
	case CategoryAnalysis:
		return "analysis"
	case CategoryGeneral:
		return "general"
	default:
		return "unknown"
	}
}

// RoutingRule maps a ModelCategory to a specific provider and model.
type RoutingRule struct {
	Category ModelCategory
	Provider string
	Model    string
}

// ModelRouter routes model categories to specific provider+model combinations.
type ModelRouter struct {
	rules   []RoutingRule
	defaultRule RoutingRule
}

// NewModelRouter creates a new ModelRouter with the given rules.
// If no rule matches, the first rule is used as the default.
// If no rules are provided, a built-in default is used.
func NewModelRouter(rules []RoutingRule) *ModelRouter {
	defaultRule := RoutingRule{
		Category: CategoryGeneral,
		Provider: "openai",
		Model:    "gpt-4",
	}
	if len(rules) > 0 {
		defaultRule = rules[0]
	}
	return &ModelRouter{
		rules:       rules,
		defaultRule: defaultRule,
	}
}

// Route returns the provider and model for the given category.
// If no exact match is found, the default rule is used.
func (r *ModelRouter) Route(category ModelCategory) (string, string) {
	for _, rule := range r.rules {
		if rule.Category == category {
			return rule.Provider, rule.Model
		}
	}
	return r.defaultRule.Provider, r.defaultRule.Model
}

File: packages/ai/src/router.ts
// ---------------------------------------------------------------------------
// ModelRouter
// ---------------------------------------------------------------------------

/**
 * Classifies the type of AI task being performed.
 *
 * Standard categories cover common agent operations.
 * OmO-style categories (Quick, Deep, Visual, Ultrabrain) provide
 * intent-driven routing where the user picks a mode, not a model.
 */
export enum ModelCategory {
  // Standard categories
  Coding = "coding",
  Planning = "planning",
  Review = "review",
  Answer = "answer",
  Explore = "explore",
  Creative = "creative",
  Analysis = "analysis",
  General = "general",

  // OmO-style intent categories
  /** Single-file changes, typos, quick fixes — fast + cheap model */
  Quick = "quick",
  /** Autonomous research + execution — strong reasoning model */
  Deep = "deep",
  /** Frontend, UI/UX, design — vision-capable model */
  Visual = "visual",
  /** Hard logic, architecture decisions — strongest available model */
  Ultrabrain = "ultrabrain",
}

/**
 * Maps a ModelCategory to a specific provider and model.
 */
export interface RoutingRule {
  category: ModelCategory;
  provider: string;
  model: string;
}

/**
 * Routes model categories to specific provider+model combinations.
 *
 * Supports both standard categories and OmO-style intent categories.
 * The router picks the best model for the task type automatically.
 */
export class ModelRouter {
  private readonly rules: RoutingRule[];
  private readonly defaultRule: RoutingRule;

  /**
   * Creates a new ModelRouter with the given rules.
   * If no rule matches, the first rule is used as the default.
   * If no rules are provided, a built-in default is used.
   */
  constructor(rules: RoutingRule[]) {
    const defaultRule: RoutingRule = {
      category: ModelCategory.General,
      provider: "openai",
      model: "gpt-4",
    };
    if (rules.length > 0) {
      this.defaultRule = rules[0]!;
    } else {
      this.defaultRule = defaultRule;
    }
    this.rules = rules;
  }

  /**
   * Returns the provider and model for the given category.
   * If no exact match is found, the default rule is used.
   */
  route(category: ModelCategory): { provider: string; model: string } {
    for (const rule of this.rules) {
      if (rule.category === category) {
        return { provider: rule.provider, model: rule.model };
      }
    }
    return { provider: this.defaultRule.provider, model: this.defaultRule.model };
  }
}

/**
 * Create a ModelRouter with OmO-style default category mappings.
 * Users only pick a category (quick/deep/visual/ultrabrain), the router
 * selects the appropriate provider and model.
 */
export function createOmORouter(overrides?: Partial<Record<ModelCategory, { provider: string; model: string }>>): ModelRouter {
  const defaults: RoutingRule[] = [
    { category: ModelCategory.Quick, provider: "anthropic", model: "claude-sonnet-4-6" },
    { category: ModelCategory.Deep, provider: "openai", model: "gpt-5.1" },
    { category: ModelCategory.Visual, provider: "openai", model: "gpt-5.1" },
    { category: ModelCategory.Ultrabrain, provider: "anthropic", model: "claude-opus-4-7" },
    { category: ModelCategory.Coding, provider: "anthropic", model: "claude-opus-4-7" },
    { category: ModelCategory.Planning, provider: "anthropic", model: "claude-opus-4-7" },
    { category: ModelCategory.Review, provider: "openai", model: "gpt-5.1" },
    { category: ModelCategory.Explore, provider: "anthropic", model: "claude-sonnet-4-6" },
    { category: ModelCategory.Analysis, provider: "openai", model: "gpt-5.1" },
    { category: ModelCategory.Answer, provider: "anthropic", model: "claude-sonnet-4-6" },
    { category: ModelCategory.Creative, provider: "anthropic", model: "claude-opus-4-7" },
    { category: ModelCategory.General, provider: "anthropic", model: "claude-opus-4-7" },
  ];

  if (overrides) {
    for (const rule of defaults) {
      const ov = overrides[rule.category];
      if (ov) {
        rule.provider = ov.provider;
        rule.model = ov.model;
      }
    }
  }

  return new ModelRouter(defaults);
}
